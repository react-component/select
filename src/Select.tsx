/**
 * To match accessibility requirement, we always provide an input in the component.
 * Other element will not set `tabIndex` to avoid `onBlur` sequence problem.
 * For focused select, we set `aria-live="polite"` to update the accessibility content.
 *
 * ref:
 * - keyboard: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/listbox_role#Keyboard_interactions
 *
 * New api:
 * - listHeight
 * - listItemHeight
 * - component
 *
 * Remove deprecated api:
 * - multiple
 * - tags
 * - combobox
 * - firstActiveValue
 * - dropdownMenuStyle
 * - openClassName (Not list in api)
 *
 * Update:
 * - `backfill` only support `combobox` mode
 * - `combobox` mode not support `labelInValue` since it's meaningless
 * - `getInputElement` only support `combobox` mode
 * - `onChange` return OptionData instead of ReactNode
 * - `filterOption` `onChange` `onSelect` accept OptionData instead of ReactNode
 * - `combobox` mode trigger `onChange` will get `undefined` if no `value` match in Option
 * - `combobox` mode not support `optionLabelProp`
 */

import * as React from 'react';
import { OptionsType as SelectOptionsType } from './interface';
import SelectOptionList from './OptionList';
import Option from './Option';
import OptGroup from './OptGroup';
import { convertChildrenToData as convertSelectChildrenToData } from './utils/legacyUtil';
import {
  getLabeledValue as getSelectLabeledValue,
  filterOptions as selectDefaultFilterOptions,
  isValueDisabled as isSelectValueDisabled,
  findValueOption as findSelectValueOption,
  flattenOptions,
  fillOptionsWithMissingValue,
} from './utils/valueUtil';
import generateSelector, { SelectProps, RefSelectProps } from './generate';
import { DefaultValueType, Key, SingleType } from './interface/generator';
import warningProps from './utils/warningPropsUtil';

const RefSelect = generateSelector<SelectOptionsType>({
  prefixCls: 'rc-select',
  components: {
    optionList: SelectOptionList,
  },
  convertChildrenToData: convertSelectChildrenToData,
  flattenOptions,
  getLabeledValue: getSelectLabeledValue,
  filterOptions: selectDefaultFilterOptions,
  isValueDisabled: isSelectValueDisabled,
  findValueOption: findSelectValueOption,
  warningProps,
  fillOptionsWithMissingValue,
});

export type ExportedSelectProps<
  ValueType extends DefaultValueType = DefaultValueType
> = SelectProps<SelectOptionsType, ValueType>;

/**
 * Typescript not support generic with function component,
 * we have to wrap an class component to handle this.
 */
class Select<
  VT extends Key | Key[] = Key,
  TKeyForOptions extends Key = Key
> extends React.Component<SelectProps<SelectOptionsType<TKeyForOptions, SingleType<VT>>, VT>> {
  static Option: typeof Option = Option;

  static OptGroup: typeof OptGroup = OptGroup;

  selectRef = React.createRef<RefSelectProps>();

  focus = () => {
    this.selectRef.current.focus();
  };

  blur = () => {
    this.selectRef.current.blur();
  };

  render() {
    return <RefSelect ref={this.selectRef} {...this.props} />;
  }
}

export default Select;
