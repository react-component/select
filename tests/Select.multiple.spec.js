/* eslint-disable no-undef */
import React from 'react';
import { mount } from 'enzyme';
import Select, { Option, OptGroup } from '../src';
import allowClearTest from './shared/allowClearTest';
import focusTest from './shared/focusTest';
import blurTest from './shared/blurTest';
import hoverTest from './shared/hoverTest';
import renderTest from './shared/renderTest';
import removeSelectedTest from './shared/removeSelectedTest';
import dynamicChildrenTest from './shared/dynamicChildrenTest';
import inputFilterTest from './shared/inputFilterTest';

describe('Select.multiple', () => {
  allowClearTest('multiple');
  focusTest('multiple');
  blurTest('multiple');
  hoverTest('multiple');
  renderTest('multiple');
  removeSelectedTest('multiple');
  dynamicChildrenTest('multiple');
  inputFilterTest('multiple');

  it('tokenize input', () => {
    const handleChange = jest.fn();
    const handleSelect = jest.fn();
    const wrapper = mount(
      <Select
        multiple
        optionLabelProp="children"
        tokenSeparators={[',']}
        onChange={handleChange}
        onSelect={handleSelect}
      >
        <OptGroup key="group1">
          <Option value="1">One</Option>
        </OptGroup>
        <OptGroup key="group2">
          <Option value="2">Two</Option>
        </OptGroup>
      </Select>,
    );

    const input = wrapper.find('input');
    input.instance().focus = jest.fn();

    input.simulate('change', {
      target: {
        value: 'One',
      },
    });

    input.simulate('change', {
      target: {
        value: 'One,Two,Three',
      },
    });

    input.simulate('change', {
      target: {
        value: 'One,Two',
      },
    });

    expect(handleChange).toBeCalledWith(['1', '2'], expect.anything());
    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(handleSelect).toHaveBeenCalledTimes(2);
    expect(wrapper.state().value).toEqual(['1', '2']);
    expect(
      wrapper
        .find('.rc-select-selection__choice__content')
        .at(0)
        .text(),
    ).toEqual('One');
    expect(
      wrapper
        .find('.rc-select-selection__choice__content')
        .at(1)
        .text(),
    ).toEqual('Two');
    expect(wrapper.state().inputValue).toBe('');
    expect(wrapper.state().open).toBe(false);
    expect(input.instance().focus).toBeCalled();
  });

  it('focus', () => {
    const handleFocus = jest.fn();
    const wrapper = mount(
      <Select multiple onFocus={handleFocus}>
        <Option value="1">One</Option>
        <Option value="2">Two</Option>
      </Select>,
    );
    jest.useFakeTimers();
    wrapper.find('.rc-select').simulate('focus');
    jest.runAllTimers();
    expect(handleFocus).toBeCalled();
  });

  it('OptGroup without key', () => {
    expect(() => {
      mount(
        <Select multiple defaultValue={['1']}>
          <OptGroup label="group1">
            <Option value="1">One</Option>
          </OptGroup>
          <OptGroup label="group2">
            <Option value="2">Two</Option>
          </OptGroup>
        </Select>,
      );
    }).not.toThrow();
  });

  it('allow number value', () => {
    const handleChange = jest.fn();

    const wrapper = mount(
      <Select multiple defaultValue={1} onChange={handleChange}>
        <Option value={1}>1</Option>
        <Option value={2} testprop={2}>
          2
        </Option>
      </Select>,
    );

    expect(wrapper.find('.rc-select-selection__choice__content').text()).toBe('1');

    wrapper.find('.rc-select').simulate('click');
    wrapper
      .find('MenuItem')
      .at(1)
      .simulate('click');
    expect(handleChange).toBeCalledWith(
      [1, 2],
      [
        <Option value={1}>1</Option>,
        <Option value={2} testprop={2}>
          2
        </Option>,
      ],
    );
    expect(
      wrapper
        .find('.rc-select-selection__choice__content')
        .at(1)
        .text(),
    ).toBe('2');
  });

  it('do not open when close button click', () => {
    const wrapper = mount(
      <Select multiple>
        <Option value={1}>1</Option>
        <Option value={2}>2</Option>
      </Select>,
    );
    wrapper.find('.rc-select-selection').simulate('click');
    wrapper
      .find('.rc-select-dropdown-menu-item')
      .at(0)
      .simulate('click');
    wrapper
      .find('.rc-select-dropdown-menu-item')
      .at(1)
      .simulate('click');
    wrapper.setState({ open: false });
    wrapper
      .find('.rc-select-selection__choice__remove')
      .at(0)
      .simulate('click');
    expect(wrapper.state('open')).toBe(false);
    expect(wrapper.state('value')).toEqual([2]);
  });

  it('select when item enter', () => {
    const wrapper = mount(
      <Select multiple>
        <Option value={1}>1</Option>
        <Option value={2}>2</Option>
      </Select>,
    );
    wrapper.find('.rc-select-selection').simulate('click');
    const meunItem = wrapper.find('.rc-select-dropdown-menu-item').at(1);
    // add active to meunItem
    meunItem
      .simulate('mouseenter')
      .simulate('mouseover')
      .simulate('keyDown', { keyCode: 13 });
    expect(wrapper.state('open')).toBe(true);
    expect(wrapper.state('value')).toEqual([2]);
    wrapper.unmount();
  });

  it('enter twice to cancel the selection', () => {
    const wrapper = mount(
      <Select multiple>
        <Option value={1}>1</Option>
        <Option value={2}>2</Option>
      </Select>,
    );
    wrapper.find('.rc-select-selection').simulate('click');
    const meunItem = wrapper.find('.rc-select-dropdown-menu-item').at(1);
    // add active to meunItem
    meunItem
      .simulate('mouseenter')
      .simulate('mouseover')
      .simulate('keyDown', { keyCode: 13 });
    meunItem
      .simulate('mouseenter')
      .simulate('mouseover')
      .simulate('keyDown', { keyCode: 13 });
    expect(wrapper.state('open')).toBe(true);
    expect(wrapper.state('value')).toEqual([]);
  });

  it('do not crash when children has empty', () => {
    const wrapper = mount(
      <Select multiple>
        {null}
        <Option value={1}>1</Option>
        <Option value={2}>2</Option>
      </Select>,
    );

    wrapper.find('.rc-select-selection').simulate('click');
    wrapper
      .find('.rc-select-dropdown-menu-item')
      .at(0)
      .simulate('click');

    // Do not crash
  });

  it('do not crash when value has empty string', () => {
    const wrapper = mount(
      <Select multiple value={['']}>
        <Option value={1}>1</Option>
        <Option value={2}>2</Option>
      </Select>,
    );

    expect(wrapper.find('.rc-select-selection__choice__content').length).toBe(1);
  });
});
