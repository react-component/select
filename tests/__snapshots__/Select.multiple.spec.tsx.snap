// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select.multiple allowClear renders correctly 1`] = `
<div
  class="rc-select rc-select-enabled rc-select-allow-clear"
>
  <div
    aria-autocomplete="list"
    aria-controls=""
    aria-expanded="false"
    aria-haspopup="true"
    class="rc-select-selection
            rc-select-selection--multiple"
    role="combobox"
  >
    <div
      class="rc-select-selection__rendered"
    >
      <ul>
        <li
          class="rc-select-search rc-select-search--inline"
        >
          <div
            class="rc-select-search__field__wrap"
          >
            <input
              autocomplete="off"
              class="rc-select-search__field"
              value=""
            />
            <span
              class="rc-select-search__field__mirror"
            >
               
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Select.multiple render render animation 1`] = `
<SelectTrigger
  ariaId="test-uuid"
  backfillValue=""
  defaultActiveFirstOption={true}
  dropdownMatchSelectWidth={true}
  dropdownMenuStyle={Object {}}
  dropdownRender={[Function]}
  dropdownStyle={Object {}}
  empty={false}
  inputValue=""
  multiple={true}
  onDropdownVisibleChange={[Function]}
  onMenuDeselect={[Function]}
  onMenuSelect={[Function]}
  onPopupFocus={[Function]}
  options={Array []}
  prefixCls="rc-select"
  showAction={
    Array [
      "click",
    ]
  }
  showSearch={true}
  value={Array []}
  visible={false}
>
  <div
    className="rc-select rc-select-enabled"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    onMouseOut={[Function]}
    onMouseUp={[Function]}
  >
    <div
      aria-autocomplete="list"
      aria-controls="test-uuid"
      aria-expanded={false}
      aria-haspopup="true"
      className="rc-select-selection
            rc-select-selection--multiple"
      key="selection"
      role="combobox"
    >
      <div
        className="rc-select-selection__rendered"
      >
        <Animate
          animation={Object {}}
          component="ul"
          componentProps={Object {}}
          onAppear={[Function]}
          onEnd={[Function]}
          onEnter={[Function]}
          onLeave={[Function]}
          transitionAppear={false}
          transitionEnter={true}
          transitionLeave={true}
          transitionName="slide-up"
        >
          <li
            className="rc-select-search rc-select-search--inline"
            key="__input"
          >
            <div
              className="rc-select-search__field__wrap"
            >
              <input
                autoComplete="off"
                className="rc-select-search__field"
                onChange={[Function]}
                onKeyDown={[Function]}
                value=""
              />
              <span
                className="rc-select-search__field__mirror"
              >
                 
              </span>
            </div>
          </li>
        </Animate>
      </div>
    </div>
  </div>
</SelectTrigger>
`;

exports[`Select.multiple render truncates tags by maxTagCount 1`] = `
<div
  class="rc-select rc-select-enabled"
>
  <div
    aria-autocomplete="list"
    aria-controls=""
    aria-expanded="false"
    aria-haspopup="true"
    class="rc-select-selection
            rc-select-selection--multiple"
    role="combobox"
  >
    <div
      class="rc-select-selection__rendered"
    >
      <ul>
        <li
          class="rc-select-selection__choice"
          role="presentation"
          style="user-select:none;-webkit-user-select:none"
          title="one"
          unselectable="on"
        >
          <div
            class="rc-select-selection__choice__content"
          >
            one
          </div>
          <span
            class="rc-select-selection__choice__remove"
          >
            <i
              class="rc-select-selection__choice__remove-icon"
            >
              ×
            </i>
          </span>
        </li>
        <li
          class="rc-select-selection__choice"
          role="presentation"
          style="user-select:none;-webkit-user-select:none"
          title="two"
          unselectable="on"
        >
          <div
            class="rc-select-selection__choice__content"
          >
            two
          </div>
          <span
            class="rc-select-selection__choice__remove"
          >
            <i
              class="rc-select-selection__choice__remove-icon"
            >
              ×
            </i>
          </span>
        </li>
        <li
          class="rc-select-selection__choice rc-select-selection__choice__disabled"
          role="presentation"
          style="user-select:none;-webkit-user-select:none"
          title="+ 1 ..."
          unselectable="on"
        >
          <div
            class="rc-select-selection__choice__content"
          >
            + 1 ...
          </div>
        </li>
        <li
          class="rc-select-search rc-select-search--inline"
        >
          <div
            class="rc-select-search__field__wrap"
          >
            <input
              autocomplete="off"
              class="rc-select-search__field"
              value=""
            />
            <span
              class="rc-select-search__field__mirror"
            >
               
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Select.multiple render truncates tags by maxTagCount and show maxTagPlaceholder 1`] = `
<div
  class="rc-select rc-select-enabled"
>
  <div
    aria-autocomplete="list"
    aria-controls=""
    aria-expanded="false"
    aria-haspopup="true"
    class="rc-select-selection
            rc-select-selection--multiple"
    role="combobox"
  >
    <div
      class="rc-select-selection__rendered"
    >
      <ul>
        <li
          class="rc-select-selection__choice"
          role="presentation"
          style="user-select:none;-webkit-user-select:none"
          title="one"
          unselectable="on"
        >
          <div
            class="rc-select-selection__choice__content"
          >
            one
          </div>
          <span
            class="rc-select-selection__choice__remove"
          >
            <i
              class="rc-select-selection__choice__remove-icon"
            >
              ×
            </i>
          </span>
        </li>
        <li
          class="rc-select-selection__choice"
          role="presentation"
          style="user-select:none;-webkit-user-select:none"
          title="two"
          unselectable="on"
        >
          <div
            class="rc-select-selection__choice__content"
          >
            two
          </div>
          <span
            class="rc-select-selection__choice__remove"
          >
            <i
              class="rc-select-selection__choice__remove-icon"
            >
              ×
            </i>
          </span>
        </li>
        <li
          class="rc-select-selection__choice rc-select-selection__choice__disabled"
          role="presentation"
          style="user-select:none;-webkit-user-select:none"
          title=""
          unselectable="on"
        >
          <div
            class="rc-select-selection__choice__content"
          >
            <span>
              Omitted
            </span>
          </div>
        </li>
        <li
          class="rc-select-search rc-select-search--inline"
        >
          <div
            class="rc-select-search__field__wrap"
          >
            <input
              autocomplete="off"
              class="rc-select-search__field"
              value=""
            />
            <span
              class="rc-select-search__field__mirror"
            >
               
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Select.multiple render truncates tags by maxTagCount and show maxTagPlaceholder function 1`] = `
<div
  class="rc-select rc-select-enabled"
>
  <div
    aria-autocomplete="list"
    aria-controls=""
    aria-expanded="false"
    aria-haspopup="true"
    class="rc-select-selection
            rc-select-selection--multiple"
    role="combobox"
  >
    <div
      class="rc-select-selection__rendered"
    >
      <ul>
        <li
          class="rc-select-selection__choice"
          role="presentation"
          style="user-select:none;-webkit-user-select:none"
          title="one"
          unselectable="on"
        >
          <div
            class="rc-select-selection__choice__content"
          >
            one
          </div>
          <span
            class="rc-select-selection__choice__remove"
          >
            <i
              class="rc-select-selection__choice__remove-icon"
            >
              ×
            </i>
          </span>
        </li>
        <li
          class="rc-select-selection__choice"
          role="presentation"
          style="user-select:none;-webkit-user-select:none"
          title="two"
          unselectable="on"
        >
          <div
            class="rc-select-selection__choice__content"
          >
            two
          </div>
          <span
            class="rc-select-selection__choice__remove"
          >
            <i
              class="rc-select-selection__choice__remove-icon"
            >
              ×
            </i>
          </span>
        </li>
        <li
          class="rc-select-selection__choice rc-select-selection__choice__disabled"
          role="presentation"
          style="user-select:none;-webkit-user-select:none"
          title=""
          unselectable="on"
        >
          <div
            class="rc-select-selection__choice__content"
          >
            <span>
              1 values omitted
            </span>
          </div>
        </li>
        <li
          class="rc-select-search rc-select-search--inline"
        >
          <div
            class="rc-select-search__field__wrap"
          >
            <input
              autocomplete="off"
              class="rc-select-search__field"
              value=""
            />
            <span
              class="rc-select-search__field__mirror"
            >
               
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Select.multiple render truncates tags by maxTagCount while maxTagCount is 0 1`] = `
<div
  class="rc-select rc-select-enabled"
>
  <div
    aria-autocomplete="list"
    aria-controls=""
    aria-expanded="false"
    aria-haspopup="true"
    class="rc-select-selection
            rc-select-selection--multiple"
    role="combobox"
  >
    <div
      class="rc-select-selection__rendered"
    >
      <ul>
        <li
          class="rc-select-selection__choice rc-select-selection__choice__disabled"
          role="presentation"
          style="user-select:none;-webkit-user-select:none"
          title="+ 3 ..."
          unselectable="on"
        >
          <div
            class="rc-select-selection__choice__content"
          >
            + 3 ...
          </div>
        </li>
        <li
          class="rc-select-search rc-select-search--inline"
        >
          <div
            class="rc-select-search__field__wrap"
          >
            <input
              autocomplete="off"
              class="rc-select-search__field"
              value=""
            />
            <span
              class="rc-select-search__field__mirror"
            >
               
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Select.multiple render truncates values by maxTagTextLength 1`] = `
<div
  class="rc-select rc-select-enabled"
>
  <div
    aria-autocomplete="list"
    aria-controls=""
    aria-expanded="false"
    aria-haspopup="true"
    class="rc-select-selection
            rc-select-selection--multiple"
    role="combobox"
  >
    <div
      class="rc-select-selection__rendered"
    >
      <ul>
        <li
          class="rc-select-selection__choice"
          role="presentation"
          style="user-select:none;-webkit-user-select:none"
          title="one"
          unselectable="on"
        >
          <div
            class="rc-select-selection__choice__content"
          >
            on...
          </div>
          <span
            class="rc-select-selection__choice__remove"
          >
            <i
              class="rc-select-selection__choice__remove-icon"
            >
              ×
            </i>
          </span>
        </li>
        <li
          class="rc-select-selection__choice"
          role="presentation"
          style="user-select:none;-webkit-user-select:none"
          title="two"
          unselectable="on"
        >
          <div
            class="rc-select-selection__choice__content"
          >
            tw...
          </div>
          <span
            class="rc-select-selection__choice__remove"
          >
            <i
              class="rc-select-selection__choice__remove-icon"
            >
              ×
            </i>
          </span>
        </li>
        <li
          class="rc-select-search rc-select-search--inline"
        >
          <div
            class="rc-select-search__field__wrap"
          >
            <input
              autocomplete="off"
              class="rc-select-search__field"
              value=""
            />
            <span
              class="rc-select-search__field__mirror"
            >
               
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
